IAM module: Also return user_name on create, to be more consistent.

stefanhorning opened the pull request:
##### SUMMARY
On create also return user_name as it's already done for all other actions. To make return values more predictable and usable.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
iam module

stefanhorning committed:
IAM role: Also return user_name on create, to be more consistent.

robertdebock commented:
Would it be logical if `user_name` is always returned, even when it's not changed? It might be easier to loop over the results in a next task.

stefanhorning commented:
Yes, IMHO the result structure should always be the same to make it more predictable. I think it was already returned when user already exists, hence I only added it to the create action. I will double check once more maybe.

robertdebock commented:
@stefanhorning, cool. It's [big pull request day](https://github.com/ansible/community/issues/407) today, so if you can show the test results, it's likely to get merged quickly.

webknjaz commented on lib/ansible/modules/cloud/amazon/iam.py:
```diff
@@ -728,7 +728,7 @@ def main():
                     module, iam, name, groups, been_updated, new_name)
             module.exit_json(
-                user_meta=meta, groups=user_groups, keys=keys, changed=changed)
+                user_meta=meta, groups=user_groups, user_name=meta['created_user']['user_name'], keys=keys, changed=changed)
```
It seems like it's already in user_meta key. Not sure whether it is needed.

stefanhorning replied:
Not sure what you mean. The point of this change is to make `user_name` available independent of the action performed under the hood, so plays that rely on the `user_name` key don't fail on reruns (where user is already present).

gundalow commented:
@stefanhorning Would you be able to show the test results, and look at the above comment regarding `user_meta` thanks in advance.

stefanhorning commented:
Will try to do it today or tomorrow.

stefanhorning commented:
@robertdebock the situation is indeed more fragmented. Apparently the module also allows changing the username, in which case it returns `old_user_name` and `new_user_name`, on delete it's `deleted_user`. On all other actions it was already `user_name` (except for the one I fixed here).

So I guess if we want to be sure user_name is always returned we would have to make it an additional return value for those other cases too.

P.S. I deducted that all by just looking at the code for now, haven't done test runs yet.

stefanhorning committed:
Also return user_name on delete action

stefanhorning commented:
Ok, I created a gist now with a playbook to test this module, see https://gist.github.com/stefanhorning/a948b9f63a5db8a2a38fc988080a13b2

Somehow when attempting to test the update action the module failed on me, so I commented that section for now to concentrate at the issue at hand.

So the output without my fix is as follows (module from currend devel branch):

```
PLAY [Play testing return values of the iam module] *********************************************************************************************************************************************************

TASK [Gathering Facts] **************************************************************************************************************************************************************************************
ok: [localhost]

TASK [Create user initially] ********************************************************************************************************************************************************************************
changed: [localhost]

TASK [Output of returned user_name] *************************************************************************************************************************************************************************
ok: [localhost] => {
    "iam_user_result.user_name": "VARIABLE IS NOT DEFINED!"
}

TASK [Repeat same task to ensure idempotence] ***************************************************************************************************************************************************************
ok: [localhost]

TASK [Output of returned user_name] *************************************************************************************************************************************************************************
ok: [localhost] => {
    "iam_user_result.user_name": "test-ansible-iam-module"
}

TASK [Test absent action] ***********************************************************************************************************************************************************************************
changed: [localhost]

TASK [Output of returned user_name] *************************************************************************************************************************************************************************
ok: [localhost] => {
    "iam_user_result.user_name": "VARIABLE IS NOT DEFINED!"
}

TASK [Output of returned deleted_user] **********************************************************************************************************************************************************************
ok: [localhost] => {
    "iam_user_result.deleted_user": "test-ansible-iam-module"
}

PLAY RECAP **************************************************************************************************************************************************************************************************
localhost                  : ok=8    changed=2    unreachable=0    failed=0 
```

With my fix present (this branch) is looks like this:

```
PLAY [Play testing return values of the iam module] *********************************************************************************************************************************************************

TASK [Gathering Facts] **************************************************************************************************************************************************************************************
ok: [localhost]

TASK [Create user initially] ********************************************************************************************************************************************************************************
changed: [localhost]

TASK [Output of returned user_name] *************************************************************************************************************************************************************************
ok: [localhost] => {
    "iam_user_result.user_name": "test-ansible-iam-module"
}

TASK [Repeat same task to ensure idempotence] ***************************************************************************************************************************************************************
ok: [localhost]

TASK [Output of returned user_name] *************************************************************************************************************************************************************************
ok: [localhost] => {
    "iam_user_result.user_name": "test-ansible-iam-module"
}

TASK [Test absent action] ***********************************************************************************************************************************************************************************
changed: [localhost]

TASK [Output of returned user_name] *************************************************************************************************************************************************************************
ok: [localhost] => {
    "iam_user_result.user_name": "test-ansible-iam-module"
}

TASK [Output of returned deleted_user] **********************************************************************************************************************************************************************
ok: [localhost] => {
    "iam_user_result.deleted_user": "test-ansible-iam-module"
}

PLAY RECAP **************************************************************************************************************************************************************************************************
localhost                  : ok=8    changed=2    unreachable=0    failed=0  
```

stefanhorning commented:
@robertdebock done your testing now. @gundalow also responded to the comment as you asked. 

Any comments on how to move forward with this now?

stefanhorning commented:
If there is no other feedback for now, I would recommend to just merge this.

stefanhorning commented:
@robertdebock @gundalow @defionscode Please merge

s-hertel reviewed (changes requested):
I'd really like to see the return documentation populated if the return keys are being modified. The return isn't documented for users data at all so I don't know if this is helpful for users who aren't inspecting the code - especially since there are still a couple code paths for iam_type == 'user' which do not return the user_name key, so the user_name key is still dependent on the action performed:
```
if new_name and new_path:
    module.exit_json(changed=changed, groups=user_groups, old_user_name=orig_name,
                     new_user_name=new_name, old_path=path, new_path=new_path, keys=key_list,
                     created_keys=new_key, user_meta=user_meta)
elif new_name and not new_path and not been_updated:
    module.exit_json(
        changed=changed, groups=user_groups, old_user_name=orig_name, new_user_name=new_name, keys=key_list,
        created_keys=new_key, user_meta=user_meta)
```

And in case you haven't seen it, I'd really recommend using the iam_user module (and the other new iam_ modules) since it uses boto3 and should have more consistent + documented output. The iam module tries to do too much and as a result does a number of things inconsistently/poorly as well as being difficult to maintain.

stefanhorning commented:
Thanks for your feedback. I will be looking into the other iam modules again to see  if they can fully replace this module for me. 

Will also try to update the docs...
